Support @new @variadic

cannorin opened the pull request:
Closes #5363.

This PR allows `@new` and `@variadic` to be used simultaneously.

```js
'use strict';
class Foo {
    constructor(...names) {
        this.names = names;
    }
}
let foo = new Foo("bar", "baz");
console.log(foo.names);
```

```rescript
type foo
@get external names: foo => array<string> = "names"
@new @variadic external makeFoo : array<string> => foo = "Foo"
let foo = makeFoo(["bar", "baz"])
Js.log(foo->names)
```

bobzhang commented:
Would `@raw` cover your use case? 
We would like to have diffs small for easy review

cannorin commented:
> Would `@raw` cover your use case?

`%raw` can't easily workaround this when the class is imported.

Also, it's kind of unnatural for me (and perhaps for many other users) that `@new @variadic` does not work. Docs also doesn't mention they don't work when used together.

> We would like to have diffs small for easy review

"small diffs" means no changes to `jscomp/main/builtin_cm*_datasets.ml` and `lib/**/*`, right?

bobzhang commented:
> "small diffs" means no changes to jscomp/main/builtin_cm*_datasets.ml and lib/**/*, right?

Yes, ideally no white space changes or in a separate commit.

cannorin commented:
@bobzhang done 👍

An-Tu commented on jscomp/runtime/caml_splice_call.ml:
```diff
@@ -38,6 +38,20 @@ let spliceApply : obj -> obj -> obj = [%raw{|function(fn,args){
+  for(i = 0; i < lastOne.length; ++i ){
+    applied.push(lastOne[i])
+  }
+  return new ctor(...applied)
```
Since the spread operator (...) is the ES6 feature I think it's not suitable here. Maybe something like this should be here?

```js
function fn (ctor, args) {
  var i, argLen;
  argLen = args.length
  var applied = [null]
  for(i = 0; i < argLen - 1; ++i){
    applied.push(args[i])
  }
  var lastOne = args[argLen - 1]
  for(i = 0; i < lastOne.length; ++i){
    applied.push(lastOne[i])
  }
  var F = Function.prototype.bind.apply(ctor, applied)
  return new F()
}
```

cannorin replied:
makes sense!

An-Tu commented on jscomp/runtime/caml_splice_call.ml:
```diff
@@ -38,6 +38,21 @@ let spliceApply : obj -> obj -> obj = [%raw{|function(fn,args){
+let spliceNewApply : obj -> obj -> obj = [%raw{|function (ctor,args){
+  var i, argLen;
+  argLen = args.length
+  var applied = []
```
Oh, I'm sorry, but last time I didn't draw your attention to one more change besides the spread operator.

Then we use the hack with `Function.prototype.bind.apply(fn, args)` then the first element in the `args` must be the new context for the `fn`. 
That is why the initial value for the `applied` should be
```js
var applied = [null]
```

Sorry for not writing this explicitly the first time. 😞

cannorin replied:
Ahhh, I'll fix it asap!

cannorin commented:
No pressure, but it would be great if this gets included in v10.0.

I continue to be available for resolving any reviews and change requests in the foreseeable future.

cristianoc commented:
@cannorin happy to take a look, but unsure what aspect should be checked.
Anything that one should focus on? That existing features are not affected? That all cases are covered? Code style? Others?

cannorin commented:
@cristianoc thank you!

Since the PR is about modifying `@new` so that it allows `@splice` to be used together, the main concerns would be:
* the existing `@new` codes would not break
* the coding style

cristianoc commented:
OK I'll have a go. Might ask some basic question about the codebase.
Meanwhile: CI looks red.

cristianoc commented on jscomp/core/lam_compile_external_call.ml:
```diff
@@ -30,6 +30,10 @@ let splice_fn_apply fn args =
+  E.runtime_call Js_runtime_modules.caml_splice_call "spliceNewApply"
+    [ fn; E.array Immutable args ]
+
 let splice_obj_fn_apply obj name args =
```
I guess this should be called `splice_obj_apply`.

cannorin replied:
Let's rename it then (it's not my code, though)

cristianoc commented on jscomp/core/lam_compile_external_call.ml:
```diff
@@ -30,6 +30,10 @@ let splice_fn_apply fn args =
   E.runtime_call Js_runtime_modules.caml_splice_call "spliceApply"
     [ fn; E.array Immutable args ]
 
+let splice_fn_new_apply fn args =
```
How about the name `splice_new_apply`?

cristianoc replied:
And why is it called `splice` and not `variadic`. Or are the two concepts not referring to the same mechanism?

cannorin replied:
I used `splice_fn_new_apply` because of the existing `splice_obj_fn_apply`. Let's rename it too.

and it's because `[@variadic]` is used to be called as `[@bs.splice]`, and it seems the code has not changed since then (yes, splice = variadic).

cristianoc commented on jscomp/frontend/external_ffi_types.pp.ml:
```diff
@@ -223,7 +224,7 @@ let check_ffi ?loc ffi : bool =
       upgrade (is_package_relative_path external_module_name.bundle);
       check_external_module_name external_module_name
-    | Js_new {external_module_name ;  name; scopes = _}
+    | Js_new {external_module_name ;  name; splice = _; scopes = _}
```
Not sure why unused fields are emphasised in this area of code.
Is that a reminder that they are there, but unused?

cannorin replied:
I'm not sure too but my guess is: if we add an additional member (like `splice` in this PR), it will result in a warning 9 (missing-record-field-pattern), so we can easily notice that some modification may be needed here.

cristianoc commented on jscomp/runtime/caml_splice_call.ml:
```diff
@@ -38,6 +38,21 @@ let spliceApply : obj -> obj -> obj = [%raw{|function(fn,args){
   return fn.apply(null,applied)
 }|}] 
 
+let spliceNewApply : obj -> obj -> obj = [%raw{|function (ctor,args){
```
Can't remember what this relies on.
Is the compiler going to rename `ctor` and `args` as required depending on context?

cannorin replied:
It will be emitted as is:

```rescript
let f : (int, int) => int =
  %raw(`function (a, b) { return a + b; }`)

let x = f(1, 2);
```


```js
var f = (function (a, b) { return a + b; });

var x = f(1, 2);
```

cristianoc commented on jscomp/runtime/caml_splice_call.ml:
```diff
@@ -38,6 +38,21 @@ let spliceApply : obj -> obj -> obj = [%raw{|function(fn,args){
+let spliceNewApply : obj -> obj -> obj = [%raw{|function (ctor,args){
+  var i, argLen;
+  argLen = args.length
+  var applied = [null]
```
There's a bunch of code duplication here and in the function above. I guess it's OK as there's no good way around that anyway.

Question: why is there a `null` here and not in the code above?

cannorin replied:
Refer to: https://github.com/rescript-lang/rescript-compiler/pull/5364#discussion_r882332723

cristianoc replied:
> Refer to: [#5364 (comment)](https://github.com/rescript-lang/rescript-compiler/pull/5364#discussion_r882332723)

Would you add a comment above.

cristianoc commented on jscomp/test/splice_test.ml:
```diff
@@ -37,6 +37,25 @@ let dynamic arr =
+   just makes an array with its length set to `x`,
+   so at least two parameters are needed
+*)
+external newArr : int -> int -> int array -> int array = "Array"
```
Does it test all there is to test, or are there other combinations one should think about?

cristianoc commented on jscomp/core/lam_compile_external_call.ml:
```diff
@@ -281,15 +285,24 @@ let translate_ffi (cxt : Lam_compile_context.t) arg_types
+      in
+      let fn = translate_scoped_module_val module_name fn scopes in
+      if splice then
+        let args, eff, dynamic = assemble_args_has_splice arg_types args in
```
Why was there already a function `assemble_args_has_splice` that's exactly what's needed here?

cristianoc commented on jscomp/core/lam_compile_external_call.ml:
```diff
@@ -281,15 +285,24 @@ let translate_ffi (cxt : Lam_compile_context.t) arg_types
-         E.new_ fn args)
+        | EffectCall _ | NeedValue _ -> ()
+      in
+      let fn = translate_scoped_module_val module_name fn scopes in
```
The order in which this function is called has changed.
How do we know that it does not matter?

cannorin replied:
Yes, we should better move `translate_scoped_module_val` below `assemble_args_has_splice`, since `translate_scoped_module_val` might introduce some side effects.

The original call order is:
1. `assemble_args_has_splice`
2. `translate_scoped_module_val`
3. `add_eff ... (E.new ...)`

cristianoc commented:
Also, the changelog should be updated.

cannorin commented:
applied the reviews 🚀

cannorin commented:
> Error: Conflicting attributes: Attribute found that conflicts with %@new

hmm, it seems that the modified compiler is not used during CI run?

I didn't stage the following files. Do I need to check in these files to make the CI run the modified compiler?
```
        modified:   jscomp/main/builtin_cmi_datasets.ml
        modified:   jscomp/main/builtin_cmj_datasets.ml
        modified:   jscomp/test/splice_test.js
        modified:   lib/4.06.1/unstable/js_compiler.ml
        modified:   lib/4.06.1/unstable/js_playground_compiler.ml
        modified:   lib/4.06.1/whole_compiler.ml
        modified:   lib/es6/caml_splice_call.js
        modified:   lib/js/caml_splice_call.js
```

cannorin commented:
I try to stage everything in a separate commit (so that it can be easily reverted) and see what happens

cannorin commented:
macos-arm didn't run but everything else is now green 🚀

cristianoc commented:
Sometimes macOS-arm needs a bit of encouragement.
Good to go. Ready to merge if there's nothing outstanding.
